### ISSUE TITLE
Doesn't the latest exoplayer even try to play if it goes beyond the specs defined on the device?

### ISSUE THREAD

**Issue by odkrys** [google/ExoPlayer#10267]
Hello.
I asked a question about revocation of widevine last time and I didn't get a proper answer, but anyway, Lenovo will provide OTA for widevine L3 problem, so the problem seems to be solved.

This is a question about some weird video playback happening on my Lenovo P11 Pro.

This tablet uses the Snapdragon 730G.
https://github.com/Parseus/codecinfo
According to the codecinfo app, the playback capability of this device is stated as 4k: 30fps in HEVC and VP9.

However, as a result of testing, I can see that HEVC/VP9 10bit 4K 60fps HDR video runs well without frame drops.
It looks like this device probably has enough decoding margin.

I have the following problems.

1. From YouTube 17.x version, when I select 4K 60fps HDR video, it plays as 480p SDR video. No problem with 4K 60fps SDR / 1440p 60fps HDR / 4k 30fps HDR video.
In 16.x version, 4k 60fps HDR video plays well.

2. In Just Player app, 4k 60fps SDR/HDR video is played in mkv format, but not in mp4 format.
https://github.com/moneytoo/Player/issues/308

I suspect this is caused by a mismatch between the exoplayer and the playback capabilities defined in the device's media_codecs.xml and the actual playback capabilities of the hardware.

In summary, 4k 60fps HDR video won't play on 17.x YouTube.
In Just Player, 4k 60fps SDR/HDR video plays in mkv format but not in mp4 format.

Are these issues related to exoplayer?

**Comment by icbaker** [comment]
In order for us to help you effectively you need to phrase your question in terms of the ExoPlayer library APIs and behaviors. We can't answer general questions about the behavior of apps that use ExoPlayer because they might have made any number of modifications to the library that we don't know the details of.

You can try playing the content using our demo app, and provide the result of `adb bugreport` taken after playback, then it will be much easier for us to discuss in concrete terms how the library is behaving (because we know how the demo app uses the library): https://exoplayer.dev/demo-application.html

----

To answer the title and ignore the content of the question:

ExoPlayer will try and play content that 'exceeds' the stated spec of a codec - sometimes it will succeed and sometimes it will fail. You can see how we decide if content exceeds a decoder's capabilities or isn't supported at all here in `MediaCodecVideoRenderer`:
https://github.com/google/ExoPlayer/blob/a308c691d120e0b8496b2e843635edd9cc7dcef0/library/core/src/main/java/com/google/android/exoplayer2/video/MediaCodecVideoRenderer.java#L375-L390

**Comment by odkrys** [comment]
I installed android studio and played the 4K 60fps H.264 sample in the demo app.
I could see a message about NO_EXCEEDS_CAPABILITIES and I don't know how to play a local file or youtube video.

The exoplayer seems to check the playability of the video a bit more, which seems to be quite different from what I want "just play and then stutter when it lacks capability".

It's unclear if the issue on YouTube is related to the exoplayer version, but the issue with 480p SDR playing when 4k 60fps HDR is selected is pretty annoying.

**Comment by odkrys** [comment]
@icbaker [icbaker](https://github.com/icbaker)
Hello.
I was just using the device as it works anyway.
However, starting today, I'm having trouble playing 4k 60fps SDR youtube videos on my device.

I posted a related thread(with video) on the Lenovo forum, but I honestly doubt if the issue can be patched by the manufacturer.
https://forums.lenovo.com/t5/Lenovo-Android-based-Tablets-and-Phablets/P11-Pro-Unable-to-play-4k-60fps-video-on-YouTube/m-p/5148300

I know google doesn't like "youtube vanced", but I know that older versions of this 3rd party app have the feature to disable exoplayer v2.
I made it to use exoplayer v1 and I can see all 4k 60fps SDR/HDR playing on my device.
(Although the high bitrate 67mbps 4k 60fps video that has just been uploaded to youtube has stuttering, most compressed 25-35mbps videos play well.)

I wonder if any compatibility patch can be made for this on youtube or exoplayer side.

**Comment by icbaker** [comment]
Closing this issue, because the question about how ExoPlayer decides what's playable or not has been answered, and the remaining questions (about the behaviour of the YouTube app) are off-topic for this issue tracker.
